File: 1058-lexicographically-smallest-equivalent-string/1058-lexicographically-smallest-equivalent-string.cpp

class Solution {
public:
    unordered_map<char, char> parent;

    string smallestEquivalentString(string s1, string s2, string baseStr) {
        // Initialize all 26 lowercase characters
        for (char ch = 'a'; ch <= 'z'; ch++) {
            parent[ch] = ch;
        }

        // Union the equivalences
        for (int i = 0; i < s1.size(); i++) {
            Union(s1[i], s2[i]);
        }

        // Build result from baseStr
        string res;
        for (int i = 0; i < baseStr.size(); i++) {
            char ch = find(baseStr[i]);
            res.push_back(ch);
        }
        return res;
    }

    char find(char ch) {
        if (parent[ch] == ch) {
            return ch;
        }
        return parent[ch] = find(parent[ch]);
    }

    void Union(char ch1, char ch2) {
        char p1 = find(ch1), p2 = find(ch2);
        if (p1 == p2) {
            return;
        }
        // Always keep the lexicographically smaller as parent
        else if (p1 < p2) {
            parent[p2] = p1;
        } else {
            parent[p1] = p2;
        }
    }
};

File: 1058-lexicographically-smallest-equivalent-string/1058-lexicographically-smallest-equivalent-string.py
class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        parent = {}
        n = len(s1)
        for i in range(n):
            parent[s1[i]] = s1[i]
            parent[s2[i]] = s2[i]

        def find(ch):
            if parent[ch] == ch:
                return ch
            parent[ch] = find(parent[ch])
            return parent[ch]

        def union(ch1, ch2):
            p1, p2 = find(ch1), find(ch2)
            if p1 == p2:
                return

            if p1 < p2:
                parent[p2] = p1
            else:
                parent[p1] = p2

        for i in range(n):
            union(s1[i], s2[i])

        ans = ""
        for ch in baseStr:
            if ch in parent:
                ans += find(ch)
            else:
                ans += ch

        return ans
